### PR TITLE
Zy/602 model add risk plot

### LINKED ISSUE #602: MODEL: Add risk plot to predictive model
# Description
Add image of risk chart using tuds for models

- [x] linear
- [x]  neural
- [x]  tree
- [x]  forest
- [x]  svm
- [x]  boost

### PR DESCRIPTION
**Pull request by zheyxu**
# Pull Request Details

## What issue does this PR address

- MODEL: Add risk plot to predictive model

- Link to associated issue: #602 

## Checklist

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [x] Screenshots included in linked issue
- [x] Changes adhere to the style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [x] Pre-exisiting lint errors noted: [HERE]
- [ ] Integration test `make qtest.tmp` screenshot included in issue
- [x] Tested on device:
  - [ ] Linux
  - [x] MacOS
  - [ ] Windows
- [x] Added two reviewers

## Finalising

Once PR discussion is complete and reviewers have approved:

- [ ] Merge dev into the branch
- [ ] Resolve any conflicts
- [ ] Add one line summary into CHANGELOG.md
- [ ] Bump appropriate version number in pubspec.yaml
- [ ] Push to git repository and review
- [ ] Merge PR into dev (gjwgit)

### CODE REVIEW

**Comment by gjwgit** [comment]
I am now getting the following error for DEMO -> WEATHER -> TREE -> BUILD -> RISK CHART (not chart is shown)

```
> risk_results <- rattle::evaluateRisk(
+   predicted = as.numeric(as.factor(predicted)), 
+   actual    = as.numeric(as.factor(actual)), 
+   risks     = as.numeric(risks)
+ )
Error in data.frame(Actual = actual, Risk = as.numeric(risks), Predict = predicted) : 
  arguments imply differing number of rows: 53, 363, 255
> levels_predicted <- unique(predicted)
> levels_actual <- unique(actual)
> predicted <- as.character(predicted)
> predicted_numeric <- ifelse(predicted == levels_predicted[1], 0, 1)
> actual_numeric <- ifelse(actual == levels_actual[1], 0, 1)
> svg("/tmp/rattleLJQDIM/model_rpart_risk.svg")
> rattle::riskchart(predicted_numeric, actual_numeric, risks) +
+   labs(title="Risk Chart - Tuning Dataset") +
+   theme(plot.title = element_text(size=14))
Error in data.frame(Actual = actual, Risk = as.numeric(risks), Predict = predicted) : 
  arguments imply differing number of rows: 53, 363, 255
```

**Review by gjwgit — CHANGES_REQUESTED** [review]
Need to test that we obtain correct risk charts for weather and audit datasets. Most are failing for me now.

**Comment by zheyxu** [comment]
> Need to test that we obtain correct risk charts for weather and audit datasets. Most are failing for me now.
Screenshot 2024-11-26 at 10 53 00 am


It works for me for weather but for audit does not work..

**Comment by gjwgit** [comment]
Seems like you are using the full training dataset not the tuning dataset that should be used for the model building phase. Be sure to turn partition on.

File: assets/r/model_build_rpart.R [hunk 1]
```diff
@@ -84,11 +84,22 @@ rattle::asRules(model_rpart)
 # Extract and print rules.
 
 rules <- rattle::asRules(model_rpart)
+
+model_rpart_tu <- rpart(
+  form,
```
Comment (gjwgit) [inline assets/r/model_build_rpart.R]: No, we don't build the model on tu. We build the model on tr and then test the performance of the model on tu.

File: assets/r/model_build_rpart.R [hunk 2]
```diff
@@ -84,11 +84,22 @@ rattle::asRules(model_rpart)
...
+  control = rpart.control(usesurrogate = 0,
+                          maxsurrogate = 0, 
+                          MINSPLIT, MINBUCKET, MAXDEPTH, CP),
+  model   = TRUE)
   
 # Prepare probabilities for predictions.
 
-predicted_probs <- predict(model_rpart, 
-                           type         = "prob")
+predicted_probs <- predict(model_rpart_tu,
```
Comment (gjwgit) [inline assets/r/model_build_rpart.R]: This should be model_rpart - we don;t build a model of the tu dataset.

File: assets/r/model_build_conditional_forest.R [hunk 3]
```diff
@@ -82,3 +83,61 @@ ggplot(importance_df, aes(x = reorder(Variable, Importance), y = Importance)) +
   theme_minimal()
 
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted <- kernlab::predict(model_conditionalForest, 
+                              newdata                  = tuds,
```
Comment (gjwgit) [inline assets/r/model_build_conditional_forest.R]: Too much space before the '=' Align them after the `newdata `

File: assets/r/model_build_conditional_forest.R [hunk 4]
```diff
@@ -82,3 +83,61 @@ ggplot(importance_df, aes(x = reorder(Variable, Importance), y = Importance)) +
...
+                              type                     = "prob")[,2]
+  
+actual <- as.character(tuds[[target]])
+  
+# Create numeric risks vector.
+
+risks <- as.character(ds[[risk]])
```
Comment (gjwgit) [inline assets/r/model_build_conditional_forest.R]: This code is repeated over and over? Use the already computed values from the model_template.R

File: assets/r/model_build_ctree.R [hunk 5]
```diff
@@ -59,7 +60,40 @@ print(model_ctree)
...
-# Plot the resulting Conditional Inference Tree
+# Plot the resulting Conditional Inference Tree.
+
 svg("TEMPDIR/model_tree_ctree.svg")
 plot(model_ctree, main = paste("Conditional Inference Tree", target))
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted_probs <- predict(model_ctree,
```
Comment (gjwgit) [inline assets/r/model_build_ctree.R]: In the model template we call this pr.

File: assets/r/model_build_ctree.R [hunk 6]
```diff
@@ -59,7 +60,40 @@ print(model_ctree)
...
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted_probs <- predict(model_ctree, 
+                           newdata      = tuds, 
+                           type         = "prob")
+predicted <- apply(predicted_probs, 1, function(x) colnames(predicted_probs)[which.max(x)])
+  
+actual <- as.character(actual_tu)
```
Comment (gjwgit) [inline assets/r/model_build_ctree.R:76]: Why not use actual_tu from model_template.R

File: assets/r/model_build_linear.R [hunk 7]
```diff
@@ -68,3 +70,44 @@ svg("TEMPDIR/model_glm_diagnostic_plots.svg")
 par(mfrow = c(2, 2))
 plot(model_glm)
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted <- predict(model_glm, 
+                     newdata    = tuds,
```
Comment (gjwgit) [inline assets/r/model_build_linear.R]: Remove too much space.

File: assets/r/model_build_neural_neuralnet.R [hunk 8]
```diff
@@ -159,3 +161,43 @@ NeuralNetTools::plotnet(model_neuralnet,
                         neg_col    = "grey",
                         node_labs  = TRUE)
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted <- predict(model_neuralnet, 
+                     newdata          = tuds,)
```
Comment (gjwgit) [inline assets/r/model_build_neural_neuralnet.R]: Remove too much space.

File: assets/r/model_build_neural_neuralnet.R [hunk 9]
```diff
@@ -159,3 +161,43 @@ NeuralNetTools::plotnet(model_neuralnet,
...
+  
+actual <- as.character(tuds[[target]])
+  
+# Create numeric risks vector.
```
Comment (gjwgit) [inline assets/r/model_build_neural_neuralnet.R]: Use risk_tu and actual_tu?

File: assets/r/model_build_random_forest.R [hunk 10]
```diff
@@ -227,3 +229,60 @@ if (min_class_size >= 3 && length(unique(predicted_probs)) > 1) {
 }
 
 dev.off()
+
+# Prepare probabilities for predictions.
+
+predicted <- kernlab::predict(model_randomForest, 
+                              newdata             = tuds,
+                              type                = "prob")[,2]
```
Comment (gjwgit) [inline assets/r/model_build_random_forest.R]: Too much space

File: assets/r/model_build_ctree.R (hunk 5 above)
Comment (zheyxu) [reply assets/r/model_build_ctree.R]: Have not find pr in model_template.R

Comment (gjwgit) [reply assets/r/model_build_ctree.R]: Thanks - I probably need to review how closely model_template.R follows the model template from my Essentials book. We can look to update model_template.R from the common code that is emerging from your implementations across the models.

**Pull request merged**